### PR TITLE
Haohui - fix permission for edit tasks

### PR DESCRIPTION
**Pull request by Lin1404**
# Description
image

## How to test:
1. check into current branch
2. do npm run build and npm run start to run this PR locally

**Permissions assign to individuals**

3. log as owner user call user A
4. go to dashboard→ Other Links→ Permissions Management→ Manage User Permissions
5. input another user name, call user B
6. find the “Edit Task”, click the add button
7. scroll down and click submit button
8. log out and login as user B(if already logged in, make sure to log out and log back in)
9. go to “Dashboard”→ "Other Links" -> "Projects" -> "WBS Button" -> "Choose any WBS" -> "Edit" -> "Edit" 
10. make some change of the task then submit.

**Managers account**

11. log as owner user call user A
12. go to dashboard→ Other Links→ Permissions Management→ Manage User Permissions
13. input manager user name, call user C
14. click the "Reset to Default" button on top right
15. scroll down and click submit button **("Edit Task" is selected by default)**
16. log out and login as user C(if already logged in, make sure to log out and log back in)
17. go to “Dashboard”→ "Other Links" -> "Projects" -> "WBS Button" -> "Choose any WBS" -> "Edit" -> "Edit" 
18. make some change of the task then submit.

## Screenshots or videos of changes:
Managers account

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/78375725/78c13707-988b-4b66-a839-c6533726787b

Permissions assign to individuals

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/78375725/3f3bdef1-67fc-4a79-a36e-610be3fe38cc

## Note:
Include the information the reviewers need to know.

### CODE REVIEW

**Review by AriaYu927 — APPROVED** [review]
Hi @Lin1404 , I've tested your PR and everything works as expected.

For individual who is assigned the permission:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/94303659/cc55499e-312b-421e-a881-e3b2bf075096


For Manger user:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/94303659/cfbb9872-753f-476a-b345-6734752482f0

**Review by ljrirene — CHANGES_REQUESTED** [review]
Hi Haohui, I logged in as a volunteer user and a manager user. 
Everything works as expected when I logged in as a volunteer user:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/34266880/84018ee4-f57c-4f11-9472-533b9f7e78b9

However, when I logged in as a manager user, I can't unselect the edit button here:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/34266880/84546b76-3ab9-449d-9e78-068f1f0fdc44

**Comment by mianmiantea2019** [comment]
Hi @Lin1404 , 

I tested with different user roles: an owner user, a manager user, and a volunteer user. Everything operated smoothly without any issues.

For Manger user:

image

image

image

For Owner user:

image

image

For individual who is assigned the permission:

image

image

image

**Comment by Lin1404** [comment]
> Hi Haohui, I logged in as a volunteer user and a manager user. Everything works as expected when I logged in as a volunteer user:
> 
>  Screen.Recording.2023-12-23.at.16.28.45.mov 
> However, when I logged in as a manager user, I can't unselect the edit button here:
> 
>  Screen.Recording.2023-12-23.at.16.20.20.mov

Hi @ljrirene , we don't want to remove this for manager user. It is a default ability for manager account.

**Review by ilyaflaks — CHANGES_REQUESTED** [review]
I tested this PR and my finding are essentially the same as those of @ljrirene 
1) checked out this branch, ran npm run build and npm run start, cleared cash. Logged in as owner. Owner can "Add" the Edit Task permission to a Volunteer account: 
PR1742-owner-to-volunt
2) Logged out, logged back in as the Volunteer who just got permissions updated. Was able to edit projects and save: 
PR1742-volunt-edit-proj
PR1742-volunt-save-proj

3) Logged out, then back in as Owner again, selected a manager account. Clicked "Reset to Default": 
PR1742-owner-edit-manager
4) This is where things diverge. Your test instructions say "15. scroll down and click submit button (double check "Edit Task" is unselected)" but "Edit Task" is enabled and the "delete" is disabled even for an Owner
PR1742-owner-can-not-disable
5) Manager can edit and save a task:
PR1742-manager-edited-task

Since the last line in your instructions is "18. make some change of the task then submit.", could it be you did not mean to disable the edit feature for the Manager account? If it was unselected and then saved, a Manager would not be able to edit the task. 
If this is the intended behavior, please comment and I'll give my approval. 
Also, this is possibly outside of the scope of this PR, but I'd like to suggest a UI change: when one clicks "Reset to Default", nothing noticeable happens. A user might think it didn't work. Perhaps add another "toast" or disable the button after it performed it's function.

**Review by Alforoan — CHANGES_REQUESTED** [review]
As you can see the volunteer user permission change worked, but the manager's edit permission is disabled.



https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/118868270/e9a583fe-284b-414f-b0e1-356982cc2300

**Comment by Lin1404** [comment]
> Since the last line in your instructions is "18. make some change of the task then submit.", could it be you did not mean to disable the edit feature for the Manager account? If it was unselected and then saved, a Manager would not be able to edit the task. If this is the intended behavior, please comment and I'll give my approval. Also, this is possibly outside of the scope of this PR, but I'd like to suggest a UI change: when one clicks "Reset to Default", nothing noticeable happens. A user might think it didn't work. Perhaps add another "toast" or disable the button after it performed it's function.

@ilyaflaks Thank you for the review. Yes, since "edit task" is a default ability for manager account, so we want to keep it in all manager account. And thanks for the "disable button" advice, I will talk with Jae about this.

**Comment by Lin1404** [comment]
> As you can see the volunteer user permission change worked, but the manager's edit permission is disabled.
> 
>  pr.review.1742.mp4

Thank you for the review. We don't want to remove this for manager user. It is a default ability for manager account.

**Review by ilyaflaks — APPROVED** [review]

**Comment by ilyaflaks** [comment]
Changing my review to an approval because Haohui clarified that the "bug" I pointed out is intentional

**Review by GKD890 — APPROVED** [review]
It works as intended, I can't load Dashboard, thus I used WBS url to test. The Manager role should by default has `Edit` ability to the task. The volunteer account before I give the permission can only `suggest` to certain task.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/57503876/88868c91-1a44-471f-8a6f-6a7d9c9bd6db

**Comment by Haoxiang310** [comment]
I've tested through owner role, volunteer role and manager role. 
For owner editing a volunteer's permission, it works as expected:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/135764868/d88cc335-27b4-42c7-a7d8-59ebabdd67b9
For manager, most of the functions works as expected, but the edit task button is disabled:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/135764868/36507d43-0787-473a-996a-09e6760584d5

**Comment by DragonSlayer2401** [comment]
Everything is working as intended. Please watch the overview video below for more details;


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/102876103/e160665b-820c-4629-a869-c30bc1600689

**Review by Sophie-lei — APPROVED** [review]
I just tasted this PR with the user and manager role. Both of them can edit the task and reload with the correct update. I believe it working as expected. Here is the video to check.
For owner role change the user role permission

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/70113822/5d37abc8-57a1-4c4b-b675-aa91ebccc6a8

For owner role change  manager role task permission 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/70113822/ea8dc6eb-eb30-49d0-a10f-de0d8d965665

**Comment by Lin1404** [comment]
> I've tested through owner role, volunteer role and manager role. For owner editing a volunteer's permission, it works as expected:
> 
>  PR6-.1742.permission.mov 
> For manager, most of the functions works as expected, but the edit task button is disabled:
> 
>  PR6-.1742.manager.mov

Hi @Haoxiang310 Thank you for the review. We don't want to remove this for manager user. It is a default ability for manager account.

**Review by pshereen — APPROVED** [review]
Hi Haohui,
I logged in as an owner user and enabled the permission to edit task for a volunteer user.

Screenshot 2024-01-18 194350

Then I logged in as a volunteer user and was able to edit the task.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/7041610/cd82d03d-8e08-40d7-9792-333b8c1f4372

Later I logged in as an owner user and set the permissions of manager user to default by clicking "Reset to Default" button. 

Screenshot 2024-01-18 194512

Then I logged in as the manager user and was able to edit the task.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/7041610/2e6240e2-4433-474a-ab3c-66f203c3aeb3

However, I have one suggestion. After clicking "Reset to Default" in manage permissions page, there is no way to let the user know that permissions are set to default. It will be good if a pop-up message is shown like the one showing after clicking submit button. I had to click two times to make sure the button was clicked, and it reset the permissions.

**Comment by xmyoot** [comment]
Logged in a Admin and set new role for test user, added permissions to edit tasks and delete tasks. Logged in as Test User with elevated permissions and was able to successfully perform both operations, please see video for more details

Uploading Screen Recording 2024-01-20 at 2.11.48 PM.mov…

**Review by KevHnh — APPROVED** [review]
Tested with an owner account and works as expected. Only suggestion which is unrelated to the functionality is that the permissions list could be collapsible (you can see me struggle to look for the right permission in the recording). Again, unrelated so great work on the fix. 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/83934825/90e13db6-71b3-4ad0-97a7-d6b2108a105b

**Comment by galgalinidhi** [comment]
Hi @Lin1404, I tested the PR for both Volunteer as well as Manager accounts. It works as expected!

For Volunteer-


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44816413/5e57fbb3-37de-41f2-b00c-2c646e6ab407

For Manager-


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44816413/187ee395-3cc1-49c5-a0ad-f46eb1cd48c8

**Review by X1aoZhang99 — APPROVED** [review]
Hi Haohui,

I've tested this PR with the volunteer role and the manager role, and they both worked as expected. 

Volunteer: 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/113070229/227d3d52-e428-40ad-870f-1af8afabc76b

Manager: 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/113070229/646a1a2a-fe30-4d96-8864-e62f085d4213

**Review by Renanluizssx — APPROVED** [review]
Hi @Lin1404, i tested your application both manager and volunteer, and they are working very good.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/91035340/c3ff544e-284e-46b2-bacb-f1348af91650

**Review by SRamen1999 — APPROVED** [review]
This PR works as intended. I used an owner account to give a volunteer account permission to edit tasks for projects. One improvement I suggest is to make the permission list collapsible so it is easier and cleaner to find certain permissions. 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/56932664/1b4ae46d-be49-4275-ab5f-3f9397093461

File: src/components/Header/Header.jsx [hunk 1]
```diff
@@ -62,7 +62,7 @@ export const Header = props => {
   const canSeeProjectManagementTab = props.hasPermission('seeProjectManagement') || props.hasPermission('seeProjectManagementTab');
   const canPostProject = props.hasPermission('postProject');
   // Tasks
-  const canUpdateTask = props.hasPermission('updateTask');
+  const canUpdateTask = props.hasPermission('updateTask') || props.auth.user?.permissions?.frontPermissions.includes('updateTask');
```
Comment (nathanah) [inline src/components/Header/Header.jsx:65]:
```suggestion
  const canUpdateTask = props.hasPermission('updateTask');
```
Invividual permissions are checked through `props.hasPermission()` in addition to role permissions.

**Review by nathanah — CHANGES_REQUESTED** [review]

**Comment by one-community** [comment]
Thank you all, moving to final review.

**Review by navneeeth — APPROVED** [review]
Hello, @Lin1404!

I reviewed your PR for Volunteer and Manager roles and the implementation works as expected. I appreciate that you have maintained the 'Edit User' permission set to default for a Manager role. 


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44264834/b6d6f868-6e15-478f-94af-b14128b57866


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44264834/0adc0ea1-a150-4799-9575-c2baf1f338d3



I am approving this PR as it is ready to commit once you address @nathanah's requested change in the canUpdateTask permission check. Thank you!

**Pull request merged**